tests/components/unifiprotect/utils.py

```diff
@@ -0,0 +1,172 @@
+
+    return updated_entity
+
+
+def assert_entity_counts(
+    hass: HomeAssistant, platform: Platform, total: int, enabled: int
+) -> None:
+    """Assert entity counts for a given platform."""
+
+    entity_registry = er.async_get(hass)
+
+    entities = [
+        e for e in entity_registry.entities if split_entity_id(e)[0] == platform.value
+    ]
+
+    print(len(entities), entities)
+    assert len(entities) == total
+    print(len(hass.states.async_all(platform.value)))
+    assert len(hass.states.async_all(platform.value)) == enabled
+
+
+def ids_from_device_description(
+    platform: Platform,
+    device: ProtectAdoptableDeviceModel,
+    description: EntityDescription,
+) -> tuple[str, str]:
+    """Return expected unique_id and entity_id for a give platform/device/description combination."""
+
+    entity_name = (
+        device.display_name.lower().replace(":", "").replace(" ", "_").replace("-", "_")
+    )
+    description_entity_name = (
+        str(description.name)
+        .lower()
+        .replace(":", "")
+        .replace(" ", "_")
+        .replace("-", "_")
+    )
+
+    unique_id = f"{device.mac}_{description.key}"
+    entity_id = f"{platform.value}.{entity_name}_{description_entity_name}"
+
+    return unique_id, entity_id
+
+
+def generate_random_ids() -> tuple[str, str]:
+    """Generate random IDs for device."""
+
+    return random_hex(24).lower(), random_hex(12).upper()
+
+
+def regenerate_device_ids(device: ProtectAdoptableDeviceModel) -> None:
+    """Regenerate the IDs on UFP device."""
+
+    device.id, device.mac = generate_random_ids()
+
+
+def add_device_ref(bootstrap: Bootstrap, device: ProtectAdoptableDeviceModel) -> None:
+    """Manually add device ref to bootstrap for lookup."""
+
+    ref = ProtectDeviceRef(id=device.id, model=device.model)
+    bootstrap.id_lookup[device.id] = ref
+    bootstrap.mac_lookup[device.mac.lower()] = ref
+
+
+def add_device(
+    bootstrap: Bootstrap, device: ProtectAdoptableDeviceModel, regenerate_ids: bool
+) -> None:
+    """Add test device to bootstrap."""
+
+    if device.model is None:
+        return
+
+    device._api = bootstrap.api
+    if isinstance(device, Camera):
+        for channel in device.channels:
+            channel._api = bootstrap.api
+
+    if regenerate_ids:
+        regenerate_device_ids(device)
+    device._initial_data = device.dict()
+
+    devices = getattr(bootstrap, f"{device.model.value}s")
+    devices[device.id] = device
+    add_device_ref(bootstrap, device)
+
+
+async def init_entry(
+    hass: HomeAssistant,
+    ufp: MockEntityFixture,
+    devices: Sequence[ProtectAdoptableDeviceModel],
+    regenerate_ids: bool = True,
+) -> None:
+    """Initialize Protect entry with given devices."""
+
+    reset_objects(ufp.api.bootstrap)
+    for device in devices:
+        add_device(ufp.api.bootstrap, device, regenerate_ids)
+
+    await hass.config_entries.async_setup(ufp.entry.entry_id)
+    await hass.async_block_till_done()
```

Comment:
Basically, all of the prvious fixtures are replacied by these 3 methods as well as the base fixtures in `conftest.py`

Some individual tests override properties that are more minor enough not to justify a whole new fixture at the start of the test before `init_entry` is called. But everything else is basically done in `init_entry` now.